### PR TITLE
Handle subscriptions w/ deleted topics cleanly.

### PR DESCRIPTION
**Pull request by tseaver**
Problem observed when trying to clean up subscriptions after failing system tests:  the API returns '_deleted-topic_' for the topic path, which prevented us from unmarshalling the subscription.

We can't use `None` for the topic, because the subscription delegates some of its behavior through its topic.

### CODE REVIEW

**Comment by tseaver** [comment]
@dhermes PTAL

File: gcloud/pubsub/subscription.py [hunk 1]
```diff
@@ -67,12 +67,16 @@ def from_api_repr(cls, resource, client, topics=None):
...
         topic_path = resource['topic']
-        topic = topics.get(topic_path)
-        if topic is None:
-            # NOTE: This duplicates behavior from Topic.from_api_repr to avoid
-            #       an import cycle.
-            topic_name = topic_name_from_path(topic_path, client.project)
-            topic = topics[topic_path] = client.topic(topic_name)
+        if topic_path == '_deleted-topic_':
+            # Use a name which cannot conflict ('#' not allowed).
+            topic = client.topic('###DELETED-TOPIC###')
```
Comment (dhermes) [inline gcloud/pubsub/subscription.py]: This is bad.

File: gcloud/pubsub/subscription.py [hunk 2]
```diff
@@ -67,12 +67,16 @@ def from_api_repr(cls, resource, client, topics=None):
         if topics is None:
             topics = {}
         topic_path = resource['topic']
-        topic = topics.get(topic_path)
-        if topic is None:
-            # NOTE: This duplicates behavior from Topic.from_api_repr to avoid
-            #       an import cycle.
-            topic_name = topic_name_from_path(topic_path, client.project)
-            topic = topics[topic_path] = client.topic(topic_name)
+        if topic_path == '_deleted-topic_':
```
Comment (dhermes) [inline gcloud/pubsub/subscription.py]: Is this a well-defined path in the API? Make a constant in the module and leave a note?

Comment (tseaver) [inline gcloud/pubsub/subscription.py]: [API docs](https://cloud.google.com/pubsub/reference/rest/v1/projects.subscriptions#Subscription.FIELDS.topic) for `projects.subscriptions.topic` expose the constant.

220eede adds/documents a constant for that value.

File: gcloud/pubsub/subscription.py (hunk 1 above)
Comment (tseaver) [inline gcloud/pubsub/subscription.py]: c1467eb

File: gcloud/pubsub/subscription.py [hunk 3]
```diff
@@ -67,12 +75,16 @@ def from_api_repr(cls, resource, client, topics=None):
...
         topic_path = resource['topic']
-        topic = topics.get(topic_path)
-        if topic is None:
-            # NOTE: This duplicates behavior from Topic.from_api_repr to avoid
-            #       an import cycle.
-            topic_name = topic_name_from_path(topic_path, client.project)
-            topic = topics[topic_path] = client.topic(topic_name)
+        if topic_path == cls._DELETED_TOPIC_PATH:
+            # Use a name which cannot conflict ('#' not allowed).
+            topic = client.topic(None)
```
Comment (dhermes) [inline gcloud/pubsub/subscription.py]: @tseaver I guess I'm still not quite sure what the point of creating this `topic` instance in this way is?

Comment (tseaver) [inline gcloud/pubsub/subscription.py]: The `Subscription` instance delegates some behavior to its topic (finding its project and client).  We can't even *delete* a subscription instance whose `topic` doesn't return those values.

Comment (dhermes) [inline gcloud/pubsub/subscription.py]: Because of the delegation. Not for other reasons? Worth having an explicit subclass representing a deleted topic?

Comment (dhermes) [inline gcloud/pubsub/subscription.py]: Or just another container for the client?

File: gcloud/pubsub/subscription.py [hunk 4]
```diff
@@ -67,12 +75,16 @@ def from_api_repr(cls, resource, client, topics=None):
...
             topics = {}
         topic_path = resource['topic']
-        topic = topics.get(topic_path)
-        if topic is None:
-            # NOTE: This duplicates behavior from Topic.from_api_repr to avoid
-            #       an import cycle.
-            topic_name = topic_name_from_path(topic_path, client.project)
-            topic = topics[topic_path] = client.topic(topic_name)
+        if topic_path == cls._DELETED_TOPIC_PATH:
+            # Use a name which cannot conflict ('#' not allowed).
```
Comment (dhermes) [inline gcloud/pubsub/subscription.py]: This comment is stale

File: gcloud/pubsub/subscription.py (hunk 3 above)
Comment (tseaver) [inline gcloud/pubsub/subscription.py]: > Because of the delegation. Not for other reasons?

A subscription is fundamentally related to a topic:  it isn't just incidental.  E.g., when calling the `create` API, the `topic` key (which we derive from `self.topic.path`) is not optional.

> Worth having an explicit subclass representing a deleted topic?

I wouldn't invent a subclass `Topic` just for this edge case.

> Or just another container for the client?

We would need to copy out the `project` and `client` from the `topic` instance passed to `Subscription.__init__`, and then add extra arguments to allow passing those values in in the edge case that there was no actual topic.  I think we're better off just creating a brain-damaged `Topic` instance.

Comment (dhermes) [inline gcloud/pubsub/subscription.py]: It just doesn't seem like a good thing to give to a user. They have no reason to think the `Topic` instance is brain-damaged. It'd be nice to have a clearer "don't use this" signal.

The topic / subscription relationship isn't the issue here, right? It's just that the `Subscription` needs to get access to the client (and the project via the client) in **some way**.

Comment (tseaver) [inline gcloud/pubsub/subscription.py]: Unless we re-plumb `Subscription.__init__` to take the `project` and `client` directly, I don't see how, unless you think we should make the `topic` optional (*ick*):  in that case, `Subscription.__init__` would copy `project` and `client` from the `topic`, but only if passed, and `Subscription.from_api_repr` would set the private attrs directly.  I don't see a win vs the current state, cleanliness-wise.

File: gcloud/pubsub/subscription.py [hunk 5]
```diff
@@ -67,12 +75,15 @@ def from_api_repr(cls, resource, client, topics=None):
...
             topics = {}
         topic_path = resource['topic']
-        topic = topics.get(topic_path)
-        if topic is None:
-            # NOTE: This duplicates behavior from Topic.from_api_repr to avoid
-            #       an import cycle.
-            topic_name = topic_name_from_path(topic_path, client.project)
-            topic = topics[topic_path] = client.topic(topic_name)
+        if topic_path == cls._DELETED_TOPIC_PATH:
+            topic = client.topic(name=None)
```
Comment (dhermes) [inline gcloud/pubsub/subscription.py]: Re-surfacing thread here. 

----

From @tseaver 

> Unless we re-plumb `Subscription.__init__` to take the `project` and `client` directly, I don't see how, unless you think we should make the `topic` optional (*ick*):  in that case, `Subscription.__init__` would copy `project` and `client` from the `topic`, but only if passed, and `Subscription.from_api_repr` would set the private attrs directly.  I don't see a win vs the current state, cleanliness-wise.

----

I agree it muddles things a bit but if we give a user one of these brain-dead `Topic` instances we should set expectations accordingly.

(Also, we don't need to pass `project` **and** `client` since the `project` comes with the client.)

**Comment by tseaver** [comment]
@dhermes 1fe6ba1 enables passing `topic=None` to the `Subscription` constructor:  in that case (which implies the "topic has been deleted" edge case), the caller must pass an explicit `client` (otherwise forbidden).  PTAL

**Comment by dhermes** [comment]
@tseaver Can we add a system test for this?

**Comment by tseaver** [comment]
> Can we add a system test for this?

@dhermes 4dbdb75 adds one (and fixes a bug it exposed).

File: system_tests/pubsub.py [hunk 6]
```diff
@@ -207,3 +207,25 @@ def test_subscription_iam_policy(self):
...
+        token = None
+        while True:
+            subs, token = Config.CLIENT.list_subscriptions(page_token=token)
+            all_subs.extend(subs)
+            if token is None:
+                break
+
+        created = [subscription for subscription in all_subs
+                   if subscription.name == ORPHANED]
+        self.assertEqual(len(created), 1)
```
Comment (dhermes) [inline system_tests/pubsub.py]: Can you verify that `created.topic is None`?

**Comment by tseaver** [comment]
@dhermes Any remaining issues?

**Comment by dhermes** [comment]
LGTM

**Comment by dhermes** [comment]
@tseaver It seems the deleted status is eventually consistent, hence our tests have grown a bit more flakiness:

https://travis-ci.org/GoogleCloudPlatform/gcloud-python/builds/120727594
https://travis-ci.org/GoogleCloudPlatform/gcloud-python/builds/120712985
